### ISSUE TITLE
Unable to onboard onto mergerfs - unicode issues...

### ISSUE THREAD

**Issue by KyleSanderson** [trapexit/mergerfs-tools#90]
Having a hell of a time trying to get this to work. Balance doesn't respect r/o drives which is brutal so I'm using dup instead but keep hitting this with international data.

```
Traceback (most recent call last):
  File "/usr/local/bin/mergerfs.dup", line 389, in <module>
    main()
  File "/usr/local/bin/mergerfs.dup", line 333, in main
    relpath  = mergerfs_relpath(fullpath)
  File "/usr/local/bin/mergerfs.dup", line 229, in mergerfs_relpath
    attr = xattr_relpath(fullpath)
  File "/usr/local/bin/mergerfs.dup", line 190, in xattr_relpath
    return lgetxattr(fullpath,'user.mergerfs.relpath')
  File "/usr/local/bin/mergerfs.dup", line 40, in lgetxattr
    return buf.raw[0:res].decode()
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 94: invalid start byte
```

**Comment by trapexit** [comment]
It's because one of your files has a badly encoded name. Due to how python works the strings have to be decoded/encoded rather than otherwise used as is. I'll take a look.

What are you doing exactly? What do you mean "onboard"? These tools are not necessary to use mergerfs. They are basic tools to help with cleanup and different common situations.

**Comment by trapexit** [comment]
Can you try https://raw.githubusercontent.com/trapexit/mergerfs-tools/encoding/src/mergerfs.dup ?

**Comment by KyleSanderson** [comment]
I haven't had it die yet - but still onboarding.

Any idea why balance trivially seems to just straight up quit? Seems to be after a nominal amount of data (1.5T~).

**Comment by trapexit** [comment]
What do you mean "onboarding"?

It exits if there is a problem with the execution of rsync or free space percentage diff is hit. Probably should print out why it's exiting to make it clear.

**Comment by KyleSanderson** [comment]
```Traceback (most recent call last):
  File "/usr/local/bin/mergerfs.dup", line 397, in <module>
    main()
  File "/usr/local/bin/mergerfs.dup", line 327, in main
    branches  = mergerfs_branches(ctrlfile)
  File "/usr/local/bin/mergerfs.dup", line 72, in mergerfs_branches
    branches = branches.split(':')
TypeError: a bytes-like object is required, not 'str'
```

**Comment by trapexit** [comment]
Could you please explain what you mean by "onboarding". You don't need to use mergerfs.dup (or any of these tool) for anything to do with using mergerfs.

**Comment by trapexit** [comment]
And please try downloading the tool again from the link above. Made some tweaks.

**Comment by KyleSanderson** [comment]
> Could you please explain what you mean by "onboarding".

So... you start with 6 disks all with different mount and bind points just from the pure churn of a couple decades worth of data. To get the data off of these disks and onto your brand new mergerfs volumes you need to follow the following process.

1. Mount these volumes as r/o.
1. Create a mergerfs volume with a brand new disk and your existing ssd which is the rootfs for this mount.
1. Duplicate the data from the r/o mounted bind mounts to get them onto the brand new disk.
1. Remove the r/o bind mount and see that your data is beautifully on the new mergerfs volume.
1. Witness a FUSE kernel BUG so restart your system (maybe forcefully too as this results in a hard hang on shutdown) and hope that kernel 5.4.6 fixes it as there were FUSE fixes in 5.4.3 from Rubrik and similar.
1. Format the old disk with mkfs.xfs -m rmapbt=1 /dev/sd*1
1. Add the old disk to the mergerfs set...
1. Rinse-Repeat with the next bind mount.

At-least this is how I on-boarded online, there didn't seem to be any other usable way.

> Made some tweaks.

Ah cool I made the same patch. Seems to be a better time here.

**Comment by trapexit** [comment]
If you didn't care about the filesystems you could have just put all the existing drives in the pool. As for transferring to the pool you could also just mount the pool then rsync each old drive. If I understand what you've done it sounds like more trouble than was necessary.

What FUSE bug?

**Comment by KyleSanderson** [comment]
> If you didn't care about the filesystems you could have just put all the existing drives in the pool.

Paths conflicted, there was no other way to move 20T of data online from the great sprawl/types of disks I have.

I haven't tested the latest patch (and it does seem like it's skipping these files as opposed to handling them) - but I think this can be merged to master and closed if they run cleanly.

**Comment by trapexit** [comment]
> Paths conflicted

Couldn't you've just have moved the paths on the drives? Would take a lot less time compared to doing full copying between devices.

Skipping what files? If you look at the code it doesn't skip files except those which match the exclude filter.

**Comment by KyleSanderson** [comment]
> Couldn't you've just have moved the paths on the drives?

distcc cares about paths along with other workloads (k8s specifically).

> Skipping what files?

The files with what appears to be Japanese utf8 track names. I'm a bit confused how rsync handles these properly - if there's an error can we not just copy the entire directory instead of skipping files?.

I think I've figured out the fnmatch stuff - wildcards don't matter at all and it's actually whole text searching? I need to do more testing but -E '*/Music/*' not matching a path to me is brutal but understandable after reviewing the code. Balance not taking -e for execute is also problematic as it's single threaded based on that choice.

Anyway, it's getting better and these should likely be 10 separate GH issues but I digress. Would PRs be better or what's the preferred avenue?

**Comment by trapexit** [comment]
> distcc cares about paths along with other workloads (k8s specifically).

I still don't follow. You could add everything to the pool then just move the data on the underlying drives which would be much faster. Clearly you area moving data around manually... why would moving it between 2 locations on 1 drive be different from moving between 2 on 2 drives?

> The files with what appears to be Japanese utf8 track names.

Can you please give me a specific example. It is impossible for me to test otherwise.  I just ran 'dup' and 'balance' on a directory with Japanese utf8 and it worked fine. 

>  if there's an error can we not just copy the entire directory instead of skipping files?.

What tool are you talking about? The tool shouldn't be erroring. If rsync returns an error it exits because a human should investigate the problem. 

> I think I've figured out the fnmatch stuff - wildcards don't matter at all and it's actually whole text searching?

On what? It is just fnmatch. Like a shell. If you want to exclude in `balance` the whole of "/Music" then you need to use "/Music/*". It seems to work just fine on my system.

> Balance not taking -e for execute is also problematic as it's single threaded based on that choice.

I don't follow. It doesn't take an "execute" command because it is written to work on the live system. It'd need a complete rewrite to keep track of sizes, file locations, etc. What does it being single threaded have to do with a dry run behavior?

> Anyway, it's getting better and these should likely be 10 separate GH issues but I digress. Would PRs be better or what's the preferred avenue?

? What do you mean? If you are offering to enhance the tool... yes... PRs.

I still don't understand why you're using dup. It sounds like you're using dup to copy between drives in the pool rather than just using rsync or mv.

**Comment by KyleSanderson** [comment]
> You could add everything to the pool then just move the data on the underlying drives which would be much faster.

No - the path names conflict with the rootfs... and the storage requirements would mimic LogN^2. Remember these were bind mounted.
```/mnt/8T1/Web /mnt/Web 
/mnt/8T1/Backup /mnt/Backup
/mnt/8T2/Recordings /mnt/Recordings 
/mnt/8T3/Recordings /mnt/Recordings/Recordings 
/mnt/4T1/Television /mnt/Web/Television 
/mnt/4T1/Films /mnt/8T3/Films/4T1
```

mergerfs is now /mnt and after on-boarding I could move the data around then because I have the space to do it. Before I had a bundle of nearly fully filled disk.

> It sounds like you're using dup to copy between drives in the pool rather than just using rsync or mv.

Yeah - I resorted to `ls -1 | parallel rsync -avhu {} /disks/12T1` when I couldn't get dup to fully work. Wasn't how I wanted to get the job done - but it did it online while still serving files out (and since it was r/o mounted - all the new files landed on 12T1).

> On what? It is just fnmatch. Like a shell. If you want to exclude in balance the whole of "/Music" then you need to use "/Music/*". It seems to work just fine on my system.

If I do something like...

`mergerfs.dup -e /mnt/Web/Music && mergerfs.dedup /mnt -E */Music/* -e`
This seems to take out my Music directory and knock it back to one copy - which isn't expected. I want the redundancy for this specific folder to be on two disks (always - as a matter of fact).  Actually thinking about it now.... this is likely getting expanded by my shell... Will look into that - sorry for the noise.

> What tool are you talking about? The tool shouldn't be erroring. If rsync returns an error it exits because a human should investigate the problem.

> Example

`03 Aria [Featuring Medi<E6>val B<E6>bes].flac`
`04 - Tr<FC>by Trio - Carajillo (Jazzanova'"'"'s '"'"'Chant For Leo'"'"' Mix).flac`
This is what's given from my POSIX shell - source is some bizarre XNU box likely from '08~. rsync is cool as a clam - python freaks out with the surrogateescape and appears to skip.

> It doesn't take an "execute" command because it is written to work on the live system.

Yeah - this is for performance. The way it's currently setup is if we're doing a re-balance and targeting 6 disks with a balance operation - the current tool will only use 2 disks in the pool (one read one write) which leaves 4 disks completely idle (which makes the balance operation take 3x as long as it should in this specific example).

**Comment by trapexit** [comment]
If it's tokenizing the values then it's not whatever the default encoding type is. The surrogateescape is for when it doesn't understand something. Neither tool skips things it doesn't understand. It skips only based on excludes. Can you just give me the files in question? Again... both tools work fine for Japanese filenames I have.

> Yeah - this is for performance. The way it's currently setup is if we're doing a re-balance and targeting 6 disks with a balance operation - the current tool will only use 2 disks in the pool (one read one write) which leaves 4 disks completely idle (which makes the balance operation take 3x as long as it should in this specific example).

That assumes no bottlenecks which is rare but regardless the only way to do things in parallel is to scan the entire system and then decided what files to move before taking any actions. You'd need to generate file manifests for moving between any 2 drives and then ensure none of those run in parallel. It's a lot more work. These tools are just usually just one offs I needed so they are built for that purpose. Running balance on a live system is perfectly safe so speed isn't all that important for most.

**Comment by KyleSanderson** [comment]
@trapexit why was this closed and the unicode fixes not merged into master? 😢

**Comment by trapexit** [comment]
I thought I had. I just did.
